internal/servers/repository_workerauth.go

```diff
@@ -0,0 +1,843 @@
+	}
+	nodeAuth.WorkerId = result.WorkerId
+
+	err = nodeAuth.ValidateNewWorkerAuth(ctx)
+	if err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+
+	_, err = r.writer.DoTx(ctx, db.StdRetryCnt, db.ExpBackoff{}, func(read db.Reader, w db.Writer) error {
+		// Store WorkerAuth
+		if err := w.Create(ctx, &nodeAuth); err != nil {
+			return errors.Wrap(ctx, err, op)
+		}
+		// Then store cert bundles associated with this WorkerAuth
+		for _, c := range node.CertificateBundles {
+			err := r.storeWorkerCertBundle(ctx, c, nodeAuth.WorkerKeyIdentifier, w)
+			if err != nil {
+				return errors.Wrap(ctx, err, op)
+			}
+		}
+		return nil
+	},
+	)
+	if err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+
+	return nil
+}
+
+func (r *WorkerAuthRepositoryStorage) storeWorkerCertBundle(ctx context.Context, bundle *types.CertificateBundle,
+	workerKeyIdentifier string, writer db.Writer,
+) error {
+	const op = "workerauth.(WorkerAuthRepositoryStorage).storeWorkerCertBundle"
+	if bundle == nil {
+		return errors.New(ctx, errors.InvalidParameter, op, "missing CertificateBundle")
+	}
+	if workerKeyIdentifier == "" {
+		return errors.New(ctx, errors.InvalidParameter, op, "empty workerKeyIdentifier")
+	}
+	if writer == nil {
+		return errors.New(ctx, errors.InvalidParameter, op, "missing writer")
+	}
+
+	workerCertBundle := allocWorkerCertBundle()
+	bundleBytes, err := proto.Marshal(bundle)
+	if err != nil {
+		return errors.New(ctx, errors.Encode, op, "error marshaling nodetypes.CertificateBundle", errors.WithWrap(err))
+	}
+
+	// Extract serial number from CA cert
+	caCert := bundle.CaCertificateDer
+	parsedCert, err := x509.ParseCertificate(caCert)
+	if err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+	workerCertBundle.CertificatePublicKey = parsedCert.AuthorityKeyId
+	workerCertBundle.CertBundle = bundleBytes
+	workerCertBundle.WorkerKeyIdentifier = workerKeyIdentifier
+
+	err = workerCertBundle.ValidateNewWorkerCertBundle(ctx)
+	if err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+
+	err = writer.Create(ctx, &workerCertBundle)
+
+	if err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+	return nil
+}
+
+func (r *WorkerAuthRepositoryStorage) extractRootCertificate(ctx context.Context, cert *types.RootCertificate) (*RootCertificate, error) {
+	const op = "workerauth.(WorkerAuthRepositoryStorage).extractRootCertificate"
+	if cert == nil {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing RootCertificate")
+	}
+
+	rootCert := allocRootCertificate()
+
+	parsedCert, err := x509.ParseCertificate(cert.CertificateDer)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+	rootCert.SerialNumber = parsedCert.SerialNumber.Uint64()
+	rootCert.Certificate = cert.CertificateDer
+	rootCert.NotValidAfter = timestamp.New(cert.NotAfter.AsTime())
+	rootCert.NotValidBefore = timestamp.New(cert.NotBefore.AsTime())
+	rootCert.PublicKey = cert.PublicKeyPkix
+	rootCert.State = cert.Id
+	rootCert.IssuingCa = ca_id
```

Comment:
I see `ca_id` several times in the PR but I never actually see where it's set...where is this coming from?

Reply:
It's in root_certificate.go- it's always "roots"

Reply:
Doesn't need to be changed now but we don't generally use underscores in names...something like `rootsId` or `caId` would be preferable